README.md
# COMPANION-12
A suite of scripts written in python to download and archive the public posts and comments of a reddit user. 
Some of the scripts are to be run manually, others are suitable for running automatically as chron jobs (bots).

The name COMPANION-12 is taken from a series of stories being written by redditor _9MOTHER9HORSE9EYES9.

By default these scripts archive the works of that author, but can be changed to backup your own or any
other user's posts and comments. Other uses might be Stylometric Analysis, Project Collaboration, etc.


changelog.txt
V 0.2 Added INI configuration in python_2.7
V 0.1 Intial commit


python_2.7/first_run.py
#! /usr/bin/python

# first_run.py - Bleep Bloop, I am an abominatio-bot 
# Run first_run.py prior to running companion-12 or your inbox will fill up!
# This should not be run as a chron job.
# Portions taken from  Chris Braun's code make_e_book.py (MIT License)
# Ver 0.2  Rob Gambrill Jun 3,2016

import ConfigParser
import webbrowser
import itertools
import operator
import os
import errno
import textwrap
import codecs
import time
from datetime import datetime

import praw


fmt = 'MHE_%Y-%m-%d_%H:%M:%S.txt'
fmt2 = '%Y-%m-%d_%H:%M:%S UCT'


Config = ConfigParser.ConfigParser()
if not (Config.read("C-12.ini")):
     print 'INI file C-1.ini Missing'
     exit()
else:
    USER_AGENT = Config.get('Credentials','USER_AGENT')
    CLIENT_ID = Config.get('Credentials','CLIENT_ID')
    CLIENT_SECRET = Config.get('Credentials','CLIENT_SECRET')
    FOLLOW_USER = Config.get('Follow','FOLLOW_USER') 
    sc_limit=int( Config.get('Posts','NUMBER_POSTS'))
    PM_TO=''


if (not USER_AGENT or not CLIENT_ID or not CLIENT_SECRET or not FOLLOW_USER):
    print 'FATAL ERROR: Missing Script Credentials!'
    print 'Please Register your script at https://www.reddit.com/prefs/apps '
    print 'and paste your script credential info into file C-12.ini'
    print ''
    print 'See the file: Creating_and_Authorizing_Reddit_Apps.txt'
    print 'for instructions.'
    exit()
else:
     client = praw.Reddit(user_agent=USER_AGENT)

print "No Personal Messages"
try:
    client.set_oauth_app_info(client_id = CLIENT_ID,
                              client_secret = CLIENT_SECRET,
                              redirect_uri="http://127.0.0.1:65010/authorize_callback")
except:
    print"FATAL"
    exit()


print 'Collecting Posts From:'
print FOLLOW_USER

r = client.get_redditor(FOLLOW_USER)


count = 0
for index, thing in enumerate(itertools.chain(r.get_comments(limit=None), r.get_submitted(limit=None)), start=1):

    count = count + 1
    text_content = getattr(thing, 'body', getattr(thing, 'selftext', None))

    text_content= thing.permalink + "  \n" + text_content

    time_temp = thing.created_utc
    file_name = datetime.utcfromtimestamp(time_temp).strftime(fmt)
    mesg_header = "COMPANION-12: Bleep, Bloop, New post detected at " + datetime.utcfromtimestamp(time_temp).strftime(fmt2)     

    try:
        file_handle = os.open(file_name, os.O_CREAT |os.O_EXCL | os.O_WRONLY,0600)
    except OSError as e:
        if e.errno == errno.EEXIST:  # Failed as the file already exists.
            print "-skipping"
            pass
        else:  # Something unexpected went wrong so reraise the exception.
            raise
    else:  # No exception, so the file must have been created successfully.
        if(PM_TO):
            client.send_message(PM_TO, mesg_header, text_content )
            print mesg_header

        print file_name
        with os.fdopen(file_handle,"w") as file_obj:
            # Using `os.fdopen` converts the handle to an object that acts like a
            # regular Python file object, and the `with` context manager means the
            # file will be automatically closed when we're done with it.
            file_obj.write(text_content.encode('ascii','replace'))


exit()



kt421_archiver_complete.py
#!/usr/bin/python3

# Code posted to /s/9M9H9E  by KiloTango421 (THANKS!)
# License: Public Domain
# I added the oauth stuff back in a generic way and found it works great!
# R.Gambrill

 
USER_AGENT='Put Name Here'
CLIENT_ID='Put Id Here'
CLIENT_SECRET='Put Secret Here'


import io
import praw
import sys
import time
 
comment_limit = None
 
def info(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)
 
 
################################################################################
 
if len(sys.argv) < 3:
    raise Exception('Usage: %s USERNAME OUTPUT_FILE' % sys.argv[0])
 
redditor_name = sys.argv[1]
output_file   = sys.argv[2]

client = praw.Reddit(user_agent=USER_AGENT)
client.set_oauth_app_info(client_id = CLIENT_ID, client_secret= CLIENT_SECRET, redirect_uri="http://127.0.0.1:65010/authorize_callback")

 
# set up API objects
info('fetching comments by ' + redditor_name)
r = praw.Reddit(user_agent='reddit-user-comments.py')
user = r.get_redditor(redditor_name)
things = user.get_overview(limit=comment_limit)
 
# fetch data from Reddit
bodies = []
for thing in things:
    if isinstance(thing, praw.objects.Submission):
        thing_type = 'submission'
        body = thing.selftext
    else:
        thing_type = 'comment'
        body = thing.body
 
    bodies.append(': %s @ [%s](%s)***\n\n%s\n___\n' % (thing_type,
                                                       time.strftime('%Y-%m-%d %H:%M:%S UTC',
                                                                     time.gmtime(thing.created_utc)),
                                                       thing.permalink,
                                                       body.strip()))
    info('\rdownloaded %d comments' % len(bodies), end='', flush=True)
info('')
 
# write output file
info('writing ' + output_file)
with io.open(output_file, 'w', encoding='utf-8') as out:
    for n, body in enumerate(reversed(bodies)):
        print('___\n***#' + str(n+1) + body, file=out)
 
info('done!')


python_2.7/companion-12.py
#! /usr/bin/python

# Companion-12.py - Bleep Bloop, I am an abominatio-bot 
# Run first_run.py prior to running this or your inbox will fill up!
# Run as a chron job
# Portions taken from  Chris Braun's code make_e_book.py (MIT License)
# Ver 0.2  Rob Gambrill May 31,2016

import ConfigParser
import webbrowser
import itertools
import operator
import os
import errno
import textwrap
import codecs
import time
from datetime import datetime

import praw


fmt = 'MHE_%Y-%m-%d_%H:%M:%S.txt'
fmt2 = '%Y-%m-%d_%H:%M:%S UCT'


Config = ConfigParser.ConfigParser()
if not (Config.read("C-12.ini")):
     print 'INI file C-1.ini Missing'
     exit()
else:
    USER_AGENT = Config.get('Credentials','USER_AGENT')
    CLIENT_ID = Config.get('Credentials','CLIENT_ID')
    CLIENT_SECRET = Config.get('Credentials','CLIENT_SECRET')
    PM_TO = Config.get('Recipient','PM_TO')
    ACCESS_TOKEN = Config.get('Tokens','ACCESS_TOKEN')
    REFRESH_TOKEN = Config.get('Tokens','REFRESH_TOKEN')
    FOLLOW_USER = Config.get('Follow','FOLLOW_USER') 
    sc_limit=int( Config.get('Posts','NUMBER_POSTS'))


if (not USER_AGENT or not CLIENT_ID or not CLIENT_SECRET or not FOLLOW_USER):
    print 'FATAL ERROR: Missing Script Credentials!'
    print 'Please Register your script at https://www.reddit.com/prefs/apps '
    print 'and paste your script credential info into file C-12.ini'
    print ''
    print 'See the file: Creating_and_Authorizing_Reddit_Apps.txt'
    print 'for instructions.'
    exit()
else:
     client = praw.Reddit(user_agent=USER_AGENT)

if (not PM_TO or not ACCESS_TOKEN or not REFRESH_TOKEN):
    print "No Personal Messages"
    try:
        client.set_oauth_app_info(client_id = CLIENT_ID,
                                  client_secret = CLIENT_SECRET,
                                  redirect_uri="http://127.0.0.1:65010/authorize_callback")
    except:
        print"FATAL"
        exit()
else:
    from prawoauth2 import PrawOAuth2Mini
    scopes =['identity','history','privatemessages', 'submit', 'read']
    oauth_helper = PrawOAuth2Mini(client,app_key = CLIENT_ID, 
                              app_secret = CLIENT_SECRET,
                              access_token = ACCESS_TOKEN,
                              refresh_token = REFRESH_TOKEN,
                              scopes=scopes)

    print "Go for sending Personal Messages to:"
    print PM_TO
    print 
    
print 'Collecting Posts From:'
print FOLLOW_USER
print 'Max Posts:'
print sc_limit
r = client.get_redditor(FOLLOW_USER)


count = 0
for index, thing in enumerate(itertools.chain(r.get_comments(limit=sc_limit), r.get_submitted(limit=sc_limit)), start=1):

    count = count + 1
    text_content = getattr(thing, 'body', getattr(thing, 'selftext', None))

    text_content= thing.permalink + "  \n" + text_content

    time_temp = thing.created_utc
    file_name = datetime.utcfromtimestamp(time_temp).strftime(fmt)
    mesg_header = "COMPANION-12: Bleep, Bloop, New post detected at " + datetime.utcfromtimestamp(time_temp).strftime(fmt2)     

    try:
        file_handle = os.open(file_name, os.O_CREAT |os.O_EXCL | os.O_WRONLY,0600)
    except OSError as e:
        if e.errno == errno.EEXIST:  # Failed as the file already exists.
            print "-skipping"
            pass
        else:  # Something unexpected went wrong so reraise the exception.
            raise
    else:  # No exception, so the file must have been created successfully.
        if(PM_TO):
            client.send_message(PM_TO, mesg_header, text_content )
            print mesg_header

        print file_name
        with os.fdopen(file_handle,"w") as file_obj:
            # Using `os.fdopen` converts the handle to an object that acts like a
            # regular Python file object, and the `with` context manager means the
            # file will be automatically closed when we're done with it.
            file_obj.write(text_content.encode('ascii','replace'))


exit()



summonable/companion-12_summon.py
#! /usr/bin/python3

# Companion-12.py - Bleep Bloop, I am an abominatio-bot 
# This bot responds to mentions of its name in comments.
# Reply's with latest post of _9MOTHER9HORSE9EYES9
#
# Requires user permission to run. Add Access and Refresh Tokens
# to ini file.
# 
# Portions taken from  Chris Braun's code make_e_book.py (MIT License)
# Ver 0.2  Rob Gambrill May 31,2016

import configparser
import itertools
import operator
import io
import os
import re
import errno
import textwrap
import codecs
import time
from datetime import datetime
from pprint import pprint

import praw


fmt = 'MHE_%Y-%m-%d_%H:%M:%S.txt'
fmt2 = '%Y-%m-%d_%H:%M:%S UCT'


Config = configparser.ConfigParser()
if not (Config.read("C-12.ini")):
     print ('INI file C-1.ini Missing')
     exit()
else:
    USER_AGENT = Config.get('Credentials','USER_AGENT')
    CLIENT_ID = Config.get('Credentials','CLIENT_ID')
    CLIENT_SECRET = Config.get('Credentials','CLIENT_SECRET')
    PM_TO = Config.get('Recipient','PM_TO')
    ACCESS_TOKEN = Config.get('Tokens','ACCESS_TOKEN')
    REFRESH_TOKEN = Config.get('Tokens','REFRESH_TOKEN')
    FOLLOW_USER = Config.get('Follow','FOLLOW_USER') 
    sc_limit=int( Config.get('Posts','NUMBER_POSTS'))


if (not USER_AGENT or not CLIENT_ID or not CLIENT_SECRET or not FOLLOW_USER):
    print ('FATAL ERROR: Missing Script Credentials!')
    print ('Please Register your script at https://www.reddit.com/prefs/apps ')
    print ('and paste your script credential info into file C-12.ini')
    print ('')
    print ('See the file: Creating_and_Authorizing_Reddit_Apps.txt')
    print ('for instructions.')
    exit()
else:
     client = praw.Reddit(user_agent=USER_AGENT)

if (not ACCESS_TOKEN or not REFRESH_TOKEN):
    print("FATAL:Required Tokens not in C-12.ini")
    exit()
else:        
    from prawoauth2 import PrawOAuth2Mini
    scopes =['identity','history','privatemessages', 'submit', 'read']
    oauth_helper = PrawOAuth2Mini(client,app_key = CLIENT_ID, 
                              app_secret = CLIENT_SECRET,
                              access_token = ACCESS_TOKEN,
                              refresh_token = REFRESH_TOKEN,
                              scopes=scopes)

    
if not os.path.isfile("posts_replied_to.txt"):
     posts_replied_to = []
else:
     with open("posts_replied_to.txt", "r") as f:
       posts_replied_to = f.read()
       posts_replied_to = posts_replied_to.split("\n")
       posts_replied_to = filter(None, posts_replied_to)
       posts_replied_to = list (posts_replied_to)
   
mentions = list (client.get_mentions(limit=None));

for mention in mentions:
     mid =mention.id
     perm = mention.permalink
     if mid not in posts_replied_to:
          posts_replied_to.append(mid)
          r = client.get_redditor(FOLLOW_USER)

          for index, thing in enumerate(itertools.chain( r.get_submitted(limit=1)), start=1):
               last_submission= thing.permalink
               s_time=thing.created_utc
          for index, thing in enumerate(itertools.chain( r.get_comments(limit=1)), start=1):
               last_comment= thing.permalink
               c_time=thing.created_utc
          postbody= "COMPANION-12: Bleep, Bloop, I am an Abominatio-Bot!  \n  Most Recent Post of _9MOTHER9HORSE9EYES9  \n  \n"

          if s_time > c_time:
               postbody= postbody + "At " +  datetime.utcfromtimestamp(s_time).strftime(fmt2) +" The Author Submitted:  \n" + last_submission + "  \n"
              
          if c_time > s_time: 
               postbody= postbody + "At " + datetime.utcfromtimestamp(c_time).strftime(fmt2) +" The Author Commented:  \n" + last_comment + "  \n" 
         
          postbody = re.sub(r"www.reddit.com","np.reddit.com", postbody)     
          postbody= postbody + "\n  [*All beings of flesh can interface here....*](https://www.reddit.com/r/9M9H9E9/)  \n"
          print(postbody)
          mention.reply(postbody)
          mention.mark_as_read()  

print  ("Run Complete!")        
print  (posts_replied_to)
with open("posts_replied_to.txt", "w") as f:
     for post_id in posts_replied_to:
          f.write(post_id + "\n") 

exit()


standalone/sa_reddit-user-comments.py
#!/usr/bin/python3

# This is a script by reddit user /u/tangokilo421
# Public domain
#
# A friend was having trouble adding site packages
# under windows, so I modifed this script to use
# packages bundled in the zip file it comes in.
#
# Based on an idea I found at:
# https://vilimpoc.org/blog/2012/11/23/bundling-up-distributable-python-package-libraries-using-pip-and-zip/ 
#
# Modified: Jun 4,2016 by /u/Robgambrill



import sys
if not ('packages' in sys.path):
    sys.path.insert(0, 'packages')
import io
import praw
import time

comment_limit = None

def info(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)


################################################################################

if len(sys.argv) < 3:
    raise Exception('Usage: %s USERNAME OUTPUT_FILE' % sys.argv[0])

redditor_name = sys.argv[1]
output_file   = sys.argv[2]

# set up API objects
info('fetching comments by ' + redditor_name)
r = praw.Reddit(user_agent='reddit-user-comments.py')
user = r.get_redditor(redditor_name)
things = user.get_overview(limit=comment_limit)

# fetch data from Reddit
bodies = []
for thing in things:
    if isinstance(thing, praw.objects.Submission):
        thing_type = 'submission'
        body = thing.selftext
    else:
        thing_type = 'comment'
        body = thing.body

    bodies.append(': %s @ [%s](%s)***\n\n%s\n___\n' % (thing_type,
                                                       time.strftime('%Y-%m-%d %H:%M:%S UTC',
                                                                     time.gmtime(thing.created_utc)),
                                                       thing.permalink,
                                                       body.strip()))
    info('\rdownloaded %d comments' % len(bodies), end='', flush=True)
info('')

# write output file
info('writing ' + output_file)
with io.open(output_file, 'w', encoding='utf-8') as out:
    for n, body in enumerate(reversed(bodies)):
        print('___\n***#' + str(n+1) + body, file=out)

info('done!')


COMPANION-12.py
# Companion-12.py - Bleep Bloop, I am an abominatio-bot 
# Run first_run.py prior to running this or your inbox will fill up!
# Run as a chron job
# Portions taken from  Chris Braun's code make_e_book.py (MIT License)
# Ver 0.1  Rob Gambrill May 31,2016


import itertools
import operator
import os
import errno
import textwrap
import codecs
import time
from datetime import datetime


import praw

# Name of Author to watch
USERNAME = '_9MOTHER9HORSE9EYES9'

# Follow API guidelines: https://github.com/reddit/reddit/wiki/API#rules
# Get App Credentials: https://www.reddit.com/prefs/apps
# Then take the info and edit these lines below.

USER_AGENT = 'your_scipt_id'
CLIENT_ID = 'your_client_id'
CLIENT_SECRET='your_client_secret'

# Your Reddit Account info
MY_USERNAME='my_username'
MY_PASSWORD= 'my_password'

fmt = 'MHE_%Y-%m-%d_%H:%M:%S.txt'
fmt2 = '%Y-%m-%d_%H:%M:%S UCT'



def main():
    client = praw.Reddit(user_agent=USER_AGENT)
    client.set_oauth_app_info(client_id = CLIENT_ID, client_secret = CLIENT_SECRET, redirect_uri="http://127.0.0.1:65010/authorize_callback")
    #log in for messaging
    client.login(MY_USERNAME, MY_PASSWORD)
    
    r = client.get_redditor(USERNAME)
    count = 0
    for index, thing in enumerate(itertools.chain(r.get_comments(limit=None), r.get_submitted(limit=None)), start=1):
      
        count = count + 1
        content = getattr(thing, 'body_html', getattr(thing, 'selftext_html', None))
        text_content = getattr(thing, 'body', getattr(thing, 'selftext', None))
        
        text_content= thing.permalink + "  \n" + text_content
        
        time_temp = thing.created_utc
        file_name = datetime.utcfromtimestamp(time_temp).strftime(fmt)
        mesg_header = "COMPANION-12: Bleep, Bloop, New post detected at " + datetime.utcfromtimestamp(time_temp).strftime(fmt2)     

        

        try:
            file_handle = os.open(file_name, os.O_CREAT |os.O_EXCL | os.O_WRONLY,0600)
        except OSError as e:
            if e.errno == errno.EEXIST:  # Failed as the file already exists.
                print "-skipping"
                pass
            else:  # Something unexpected went wrong so reraise the exception.
                raise
        else:  # No exception, so the file must have been created successfully.
            client.send_message(MY_USERNAME, mesg_header, text_content )
            print mesg_header 
            print file_name
            with os.fdopen(file_handle,"w") as file_obj:
                # Using `os.fdopen` converts the handle to an object that acts like a
                # regular Python file object, and the `with` context manager means the
                # file will be automatically closed when we're done with it.
                file_obj.write(text_content.encode('ascii','replace'))
               



if __name__ == '__main__':
    main()
